src/io/copy.php

```diff
@@ -0,0 +1,36 @@
+<?hh
+/*
+ *  Copyright (c) 2004-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace HH\Lib\IO;
+
+use namespace HH\Lib\Str;
+
+/**
+ * Copies content from the given $source read handle to the $target write handle
+ * until the eol of $source is reached.
+ *
+ * If `$chunk_size` is `null`, there is no limit on the chunk size - this function will
+ * copy the content of $source all at once.
+ */
+async function copy(
+  <<__AcceptDisposable>> ReadHandle $source,
+  <<__AcceptDisposable>> WriteHandle $target,
+  ?int $chunk_size = null,
+  ?float $timeout_seconds = null,
+): Awaitable<void> {
+  if (!$source->isEndOfFile()) {
+    $content = await $source->readAsync($chunk_size, $timeout_seconds);
+    await $target->writeAsync($content, $timeout_seconds);
+  }
+
+  if (!$source->isEndOfFile()) {
```

Comment:
You are checking for the eof before the recursive call and as the first statement.
Couldn't we skip one of the checks by doing something like this?
```HACK
{
  if ($source->isEndOfFile()) {
    return;
  }
  $content = await $source->readAsync($chunk_size, $timeout_seconds);
  await $target->writeAsync($content, $timeout_seconds);
  await copy($source, $target, $chunk_size, $timeout_seconds);
}
```

Reply:
this will result in a redundant extra call to copy, creating an awaitable and awaiting it for no reason.

Reply:
isEndOfFile is gone, it's unclear if there's a reasonably consistent version of it. The old version would sometimes mean 'never any more data to read', and sometimes mean 'no more data to read right now'.